isso/db/__init__.py

```diff
@@ -121,3 +121,39 @@ def first(rv):
 
                 con.execute('PRAGMA user_version = 3')
                 logger.info("%i rows changed", con.total_changes)
+
+        # "text" field in "comments" became NOT NULL
+        if self.version == 3:
+            with sqlite3.connect(self.path) as con:
+                con.execute("BEGIN TRANSACTION")
+                con.execute("UPDATE comments SET text = '' WHERE text IS NULL")
+                logger.info("%i rows changed", con.total_changes)
+
+                # create new table with NOT NULL constraint for "text" field
+                con.execute('CREATE TABLE IF NOT EXISTS comments_new ('
+                            '    tid REFERENCES threads(id), id INTEGER PRIMARY KEY, parent INTEGER,'
+                            '    created FLOAT NOT NULL, modified FLOAT, mode INTEGER, remote_addr VARCHAR,'
+                            '    text VARCHAR NOT NULL, author VARCHAR, email VARCHAR, website VARCHAR,'
+                            '    likes INTEGER DEFAULT 0, dislikes INTEGER DEFAULT 0, voters BLOB NOT NULL,'
+                            '    notification INTEGER DEFAULT 0);')
+
+                try:
+                    # copy data from old table to new table
+                    con.execute("""
+                        INSERT INTO comments_new (
+                            tid, id, parent, created, modified, mode, remote_addr, text, author, email, website, likes, dislikes, voters, notification
+                        )
+                        SELECT
+                            tid, id, parent, created, modified, mode, remote_addr, text, author, email, website, likes, dislikes, voters, notification
+                        FROM comments
+                    """)
+
+                    # swap tables
+                    con.execute("ALTER TABLE comments RENAME TO comments_backup_v3")
```

Comment:
I personally don't like littering our users' database with unused tables.

This might also "break" promises to users' setups where they scrub email/IP PII from `comments` or vacuum comments older than x days. They would not expect this info to be retained and this might even create legal troubles for them.

We should rather be confident that this migration succeeds and does not destroy data. You have written tests and been cautious. If the migration fails, you auto-rollback. We can `DROP` the backup table imo.